syntax = "proto3";

option optimize_for = CODE_SIZE;

import "google/protobuf/any.proto";

package protocol;

enum ResourceCode {
    BANDWIDTH = 0x00;
    ENERGY = 0x01;
    TRON_POWER = 0x02;
}

enum AccountType {
    Normal = 0;
    AssetIssue = 1;
    Contract = 2;
}

message AccountId {
    bytes name = 1;
    bytes address = 2;
}

message authority {
    AccountId account = 1;
    bytes permission_name = 2;
}

message Permission {
    enum PermissionType {
        Owner = 0;
        Witness = 1;
        Active = 2;
    }

    PermissionType type = 1;
    int32 id = 2;
    string permission_name = 3;
    int64 threshold = 4;
    int32 parent_id = 5;
    bytes operations = 6;
    repeated Key keys = 7;
}

message MarketOrderDetail {
    bytes makerOrderId = 1;
    bytes takerOrderId = 2;
    int64 fillSellQuantity = 3;
    int64 fillBuyQuantity = 4;
}

message Key {
    bytes address = 1;
    int64 weight = 2;
}

message Transaction {
    message Contract {
        enum ContractType {
            AccountCreateContract = 0;
            TransferContract = 1;
            TransferAssetContract = 2;
            VoteAssetContract = 3;
            VoteWitnessContract = 4;
            WitnessCreateContract = 5;
            AssetIssueContract = 6;
            WitnessUpdateContract = 8;
            ParticipateAssetIssueContract = 9;
            AccountUpdateContract = 10;
            FreezeBalanceContract = 11;
            UnfreezeBalanceContract = 12;
            WithdrawBalanceContract = 13;
            UnfreezeAssetContract = 14;
            UpdateAssetContract = 15;
            ProposalCreateContract = 16;
            ProposalApproveContract = 17;
            ProposalDeleteContract = 18;
            SetAccountIdContract = 19;
            CustomContract = 20;
            CreateSmartContract = 30;
            TriggerSmartContract = 31;
            GetContract = 32;
            UpdateSettingContract = 33;
            ExchangeCreateContract = 41;
            ExchangeInjectContract = 42;
            ExchangeWithdrawContract = 43;
            ExchangeTransactionContract = 44;
            UpdateEnergyLimitContract = 45;
            AccountPermissionUpdateContract = 46;
            ClearABIContract = 48;
            UpdateBrokerageContract = 49;
            ShieldedTransferContract = 51;
            MarketSellAssetContract = 52;
            MarketCancelOrderContract = 53;
            FreezeBalanceV2Contract = 54;
            UnfreezeBalanceV2Contract = 55;
            WithdrawExpireUnfreezeContract = 56;
            DelegateResourceContract = 57;
            UnDelegateResourceContract = 58;
            CancelAllUnfreezeV2Contract = 59;
        }

        ContractType type = 1;
        google.protobuf.Any parameter = 2;
        bytes provider = 3;
        bytes ContractName = 4;
        int32 Permission_id = 5;
    }

    message Result {
        enum code {
            SUCESS = 0;
            FAILED = 1;
        }

        enum contractResult {
            DEFAULT = 0;
            SUCCESS = 1;
            REVERT = 2;
            BAD_JUMP_DESTINATION = 3;
            OUT_OF_MEMORY = 4;
            PRECOMPILED_CONTRACT = 5;
            STACK_TOO_SMALL = 6;
            STACK_TOO_LARGE = 7;
            ILLEGAL_OPERATION = 8;
            STACK_OVERFLOW = 9;
            OUT_OF_ENERGY = 10;
            OUT_OF_TIME = 11;
            JVM_STACK_OVER_FLOW = 12;
            UNKNOWN = 13;
            TRANSFER_FAILED = 14;
            INVALID_CODE = 15;
        }

        int64 fee = 1;
        code ret = 2;
        contractResult contractRet = 3;
        string assetIssueID = 14;
        int64 withdraw_amount = 15;
        int64 unfreeze_amount = 16;
        int64 exchange_received_amount = 18;
        int64 exchange_inject_another_amount = 19;
        int64 exchange_withdraw_another_amount = 20;
        int64 exchange_id = 21;
        int64 shielded_transaction_fee = 22;
        bytes orderId = 25;
        repeated MarketOrderDetail orderDetails = 26;
        int64 withdraw_expire_amount = 27;
        map<string, int64> cancel_unfreezeV2_amount = 28;
    }

    message raw {
        bytes ref_block_bytes = 1;
        int64 ref_block_num = 3;
        bytes ref_block_hash = 4;
        int64 expiration = 8;
        repeated authority auths = 9;
        bytes data = 10;
        repeated Contract contract = 11;
        bytes scripts = 12;
        int64 timestamp = 14;
        int64 fee_limit = 18;
    }

    raw raw_data = 1;
    repeated bytes signature = 2;
    repeated Result ret = 5;
}

message BlockHeader {
    message raw {
        int64 timestamp = 1;
        bytes txTrieRoot = 2;
        bytes parentHash = 3;
        bytes nonce = 5;
        bytes difficulty = 6;
        int64 number = 7;
        int64 witness_id = 8;
        bytes witness_address = 9;
        int32 version = 10;
        bytes accountStateRoot = 11;
    }

    raw raw_data = 1;
    bytes witness_signature = 2;
}

message AccountCreateContract {
    bytes owner_address = 1;
    bytes account_address = 2;
    AccountType type = 3;
}

message AccountUpdateContract {
    bytes account_name = 1;
    bytes owner_address = 2;
}

message SetAccountIdContract {
    bytes account_id = 1;
    bytes owner_address = 2;
}

message AccountPermissionUpdateContract {
    bytes owner_address = 1;
    Permission owner = 2;
    Permission witness = 3;
    repeated Permission actives = 4;
}

message AssetIssueContract {
    string id = 41;

    message FrozenSupply {
        int64 frozen_amount = 1;
        int64 frozen_days = 2;
    }
    
    bytes owner_address = 1;
    bytes name = 2;
    bytes abbr = 3;
    int64 total_supply = 4;
    repeated FrozenSupply frozen_supply = 5;
    int32 trx_num = 6;
    int32 precision = 7;
    int32 num = 8;
    int64 start_time = 9;
    int64 end_time = 10;
    int64 order = 11;
    int32 vote_score = 16;
    bytes description = 20;
    bytes url = 21;
    int64 free_asset_net_limit = 22;
    int64 public_free_asset_net_limit = 23;
    int64 public_free_asset_net_usage = 24;
    int64 public_latest_free_net_time = 25;
}

message TransferAssetContract {
    bytes asset_name = 1;
    bytes owner_address = 2;
    bytes to_address = 3;
    int64 amount = 4;
}

message UnfreezeAssetContract {
    bytes owner_address = 1;
}

message UpdateAssetContract {
    bytes owner_address = 1;
    bytes description = 2;
    bytes url = 3;
    int64 new_limit = 4;
    int64 new_public_limit = 5;
}

message ParticipateAssetIssueContract {
    bytes owner_address = 1;
    bytes to_address = 2;
    bytes asset_name = 3;
    int64 amount = 4;
}

message FreezeBalanceContract {
    bytes owner_address = 1;
    int64 frozen_balance = 2;
    int64 frozen_duration = 3;
    ResourceCode resource = 10;
    bytes receiver_address = 15;
}


message UnfreezeBalanceContract {
    bytes owner_address = 1;
    ResourceCode resource = 10;
    bytes receiver_address = 15;
}

message WithdrawBalanceContract {
    bytes owner_address = 1;
}

message TransferContract {
    bytes owner_address = 1;
    bytes to_address = 2;
    int64 amount = 3;
}

message TransactionBalanceTrace {
    message Operation {
        int64 operation_identifier = 1;
        bytes address = 2;
        int64 amount = 3;
    }

    bytes transaction_identifier = 1;
    repeated Operation operation = 2;
    string type = 3;
    string status = 4;
}

message BlockBalanceTrace {
    message BlockIdentifier {
        bytes hash = 1;
        int64 number = 2;
    }

    BlockIdentifier block_identifier = 1;
    int64 timestamp = 2;
    repeated TransactionBalanceTrace transaction_balance_trace = 3;
}

message AccountTrace {
    int64 balance = 1;
    int64 placeholder = 99;
}

message AccountIdentifier {
    bytes address = 1;
}

message AccountBalanceRequest {
    AccountIdentifier account_identifier = 1;
    BlockBalanceTrace.BlockIdentifier block_identifier = 2;
}

message AccountBalanceResponse {
    int64 balance = 1;
    BlockBalanceTrace.BlockIdentifier block_identifier = 2;
}

message FreezeBalanceV2Contract {
    bytes owner_address = 1;
    int64 frozen_balance = 2;
    ResourceCode resource = 3;
}

message UnfreezeBalanceV2Contract {
    bytes owner_address = 1;
    int64 unfreeze_balance = 2;
    ResourceCode resource = 3;
}

message WithdrawExpireUnfreezeContract {
    bytes owner_address = 1;
}

message DelegateResourceContract {
    bytes owner_address = 1;
    ResourceCode resource = 2;
    int64 balance = 3;
    bytes receiver_address = 4;
    bool lock = 5;
    int64 lock_period = 6;
}

message UnDelegateResourceContract {
    bytes owner_address = 1;
    ResourceCode resource = 2;
    int64 balance = 3;
    bytes receiver_address = 4;
}

message CancelAllUnfreezeV2Contract {
    bytes owner_address = 1;
}

message ExchangeCreateContract {
    bytes owner_address = 1;
    bytes first_token_id = 2;
    int64 first_token_balance = 3;
    bytes second_token_id = 4;
    int64 second_token_balance = 5;
}

message ExchangeInjectContract {
    bytes owner_address = 1;
    int64 exchange_id = 2;
    bytes token_id = 3;
    int64 quant = 4;
}

message ExchangeWithdrawContract {
    bytes owner_address = 1;
    int64 exchange_id = 2;
    bytes token_id = 3;
    int64 quant = 4;
}

message ExchangeTransactionContract {
    bytes owner_address = 1;
    int64 exchange_id = 2;
    bytes token_id = 3;
    int64 quant = 4;
    int64 expected = 5;
}

message MarketSellAssetContract {
    bytes owner_address = 1;
    bytes sell_token_id = 2;
    int64 sell_token_quantity = 3;
    bytes buy_token_id = 4;
    int64 buy_token_quantity = 5;
}

message MarketCancelOrderContract {
    bytes owner_address = 1;
    bytes order_id = 2;
}

message ProposalApproveContract {
    bytes owner_address = 1;
    int64 proposal_id = 2;
    bool is_add_approval = 3;
}

message ProposalCreateContract {
    bytes owner_address = 1;
    map<int64, int64> parameters = 2;
}

message ProposalDeleteContract {
    bytes owner_address = 1;
    int64 proposal_id = 2;
}
message AuthenticationPath {
    repeated bool value = 1;
}

message MerklePath {
    repeated AuthenticationPath authentication_paths = 1;
    repeated bool index = 2;
    bytes rt = 3;
}

message OutputPoint {
    bytes hash = 1;
    int32 index = 2;
}

message OutputPointInfo {
    repeated OutputPoint out_points = 1;
    int32 block_num = 2;
}

message PedersenHash {
    bytes content = 1;
}

message IncrementalMerkleTree {
    PedersenHash left = 1;
    PedersenHash right = 2;
    repeated PedersenHash parents = 3;
}

message IncrementalMerkleVoucher {
    IncrementalMerkleTree tree = 1;
    repeated PedersenHash filled = 2;
    IncrementalMerkleTree cursor = 3;
    int64 cursor_depth = 4;
    bytes rt = 5;
    OutputPoint output_point = 10;
}

message IncrementalMerkleVoucherInfo {
    repeated IncrementalMerkleVoucher vouchers = 1;
    repeated bytes paths = 2;
}

message SpendDescription {
    bytes value_commitment = 1;
    bytes anchor = 2;
    bytes nullifier = 3;
    bytes rk = 4;
    bytes zkproof = 5;
    bytes spend_authority_signature = 6;
}

message ReceiveDescription {
    bytes value_commitment = 1;
    bytes note_commitment = 2;
    bytes epk = 3;
    bytes c_enc = 4;
    bytes c_out = 5;
    bytes zkproof = 6;
}

message ShieldedTransferContract {
    bytes transparent_from_address = 1;
    int64 from_amount = 2;
    repeated SpendDescription spend_description = 3;
    repeated ReceiveDescription receive_description = 4;
    bytes binding_signature = 5;
    bytes transparent_to_address = 6;
    int64 to_amount = 7;
}

message SmartContract {
    message ABI {
        message Entry {
            enum EntryType {
                UnknownEntryType = 0;
                Constructor = 1;
                Function = 2;
                Event = 3;
                Fallback = 4;
                Receive = 5;
                Error = 6;
            }

            message Param {
                bool indexed = 1;
                string name = 2;
                string type = 3;
            }

            enum StateMutabilityType {
                UnknownMutabilityType = 0;
                Pure = 1;
                View = 2;
                Nonpayable = 3;
                Payable = 4;
            }

            bool anonymous = 1;
            bool constant = 2;
            string name = 3;
            repeated Param inputs = 4;
            repeated Param outputs = 5;
            EntryType type = 6;
            bool payable = 7;
            StateMutabilityType stateMutability = 8;
        }

        repeated Entry entrys = 1;
    }

    bytes origin_address = 1;
    bytes contract_address = 2;
    ABI abi = 3;
    bytes bytecode = 4;
    int64 call_value = 5;
    int64 consume_user_resource_percent = 6;
    string name = 7;
    int64 origin_energy_limit = 8;
    bytes code_hash = 9;
    bytes trx_hash = 10;
    int32 version = 11;
}

message ContractState {
    int64 energy_usage = 1;
    int64 energy_factor = 2;
    int64 update_cycle = 3;
}

message CreateSmartContract {
    bytes owner_address = 1;
    SmartContract new_contract = 2;
    int64 call_token_value = 3;
    int64 token_id = 4;
}

message TriggerSmartContract {
    bytes owner_address = 1;
    bytes contract_address = 2;
    int64 call_value = 3;
    bytes data = 4;
    int64 call_token_value = 5;
    int64 token_id = 6;
}

message ClearABIContract {
    bytes owner_address = 1;
    bytes contract_address = 2;
}

message UpdateSettingContract {
    bytes owner_address = 1;
    bytes contract_address = 2;
    int64 consume_user_resource_percent = 3;
}

message UpdateEnergyLimitContract {
    bytes owner_address = 1;
    bytes contract_address = 2;
    int64 origin_energy_limit = 3;
}

message SmartContractDataWrapper {
    SmartContract smart_contract = 1;
    bytes runtimecode = 2;
    ContractState contract_state = 3;
}

message BuyStorageBytesContract {
    bytes owner_address = 1;
    int64 bytes = 2;
}

message BuyStorageContract {
    bytes owner_address = 1;
    int64 quant = 2;
}

message SellStorageContract {
    bytes owner_address = 1;
    int64 storage_bytes = 2;
}

message UpdateBrokerageContract {
    bytes owner_address = 1;
    int32 brokerage = 2;
}

message VoteAssetContract {
    bytes owner_address = 1;
    repeated bytes vote_address = 2;
    bool support = 3;
    int32 count = 5;
}

message WitnessCreateContract {
    bytes owner_address = 1;
    bytes url = 2;
}

message WitnessUpdateContract {
    bytes owner_address = 1;
    bytes update_url = 12;
}

message VoteWitnessContract {
    message Vote {
        bytes vote_address = 1;
        int64 vote_count = 2;
    }

    bytes owner_address = 1;
    repeated Vote votes = 2;
    bool support = 3;
}